src/gt4py/next/common.py

```diff
@@ -84,31 +74,104 @@ def __str__(self):
         return f"{self.value}[{self.kind}]"
 
 
+class OpenBound(enum.Enum):
+    LOWER = enum.auto()
+    UPPER = enum.auto()
+
+    def __add__(self, _: int) -> Self:
+        return self
+
+    __radd__ = __add__
+
+    def __sub__(self, _: int) -> Self:
+        return self
+
+    __rsub__ = __sub__
+
+    def __le__(self, other: int | OpenBound) -> bool:
+        return self is self.LOWER or other is self.UPPER
+
+    def __lt__(self, other: int | OpenBound) -> bool:
+        return self is self.LOWER and other is not self
+
+    def __ge__(self, other: int | OpenBound) -> bool:
+        return self is self.UPPER or other is self.LOWER
+
+    def __gt__(self, other: int | OpenBound) -> bool:
+        return self is self.UPPER and other is not self
+
+
 @dataclasses.dataclass(frozen=True, init=False)
-class UnitRange(Sequence[int], Set[int]):
-    """Range from `start` to `stop` with step size one."""
+class UnitRange(Sequence[int]):
+    """
+    Range from `start` to `stop` with step size one.
 
-    start: int
-    stop: int
+    An open range is constructed by passing `None` for `start` and/or `stop`.
+    """
+
+    start: int | OpenBound
+    stop: int | OpenBound
 
-    def __init__(self, start: core_defs.IntegralScalar, stop: core_defs.IntegralScalar) -> None:
-        if start < stop:
-            object.__setattr__(self, "start", int(start))
-            object.__setattr__(self, "stop", int(stop))
+    def __init__(
+        self,
+        start: core_defs.IntegralScalar | OpenBound | None,
+        stop: core_defs.IntegralScalar | OpenBound | None,
+    ) -> None:
+        if (
+            start is None
+            or isinstance(start, OpenBound)
+            or stop is None
+            or isinstance(stop, OpenBound)
+            or start < stop
+        ):
+            if start is None:
+                start = OpenBound.LOWER
+            elif isinstance(start, OpenBound):
+                assert start is OpenBound.LOWER
```

Comment:
Combined with the suggestion above, I don't think we need to assert anything here. In my opinion the case where the wrong `OpenBound` is passed to the start/stop arg should be treated exactly the same as the case where `start > stop` in the outer `else` branch:
```suggestion
            elif start is OpenBound.LOWER:
```

Reply:
agree